stl/inc/flat_set

```diff
@@ -487,35 +487,36 @@ private:
 
     template <class _Ty>
         requires _Multi // flat_multiset
-    iterator _Emplace(_Ty&& _Val) {
+    _NODISCARD iterator _Emplace(_Ty&& _Val) {
         _STL_INTERNAL_STATIC_ASSERT(is_same_v<remove_cvref_t<_Ty>, _Kty>);
         return _Mycont.emplace(upper_bound(_Val), _STD forward<_Ty>(_Val));
     }
 
     template <class _Ty>
         requires (!_Multi) // flat_set
-    pair<iterator, bool> _Emplace(_Ty&& _Val) {
+    _NODISCARD pair<iterator, bool> _Emplace(_Ty&& _Val) {
         const iterator _Where = lower_bound(_Val);
         if (_Where != end() && !_Compare(_Val, *_Where)) {
+            // the set already contains _Val.
             return pair{_Where, false};
         }
 
         if constexpr (is_same_v<remove_cvref_t<_Ty>, _Kty>) {
             return pair{_Mycont.emplace(_Where, _STD forward<_Ty>(_Val)), true};
         } else {
             // heterogeneous insertion
-            // TRANSITION, the requirements may be stricter than needed by the standard.
             _STL_INTERNAL_STATIC_ASSERT(_Keylt_transparent && is_constructible_v<_Kty, _Ty>);
             _Kty _Keyval(_STD forward<_Ty>(_Val));
-            _STL_ASSERT(_Can_insert(_Where, _Keyval), "The conversion from the heterogeneous key to key_type should "
-                                                      "be consistent with the heterogeneous lookup!");
-            return pair{_Mycont.emplace(_Where, _STD move(_Keyval)), true};
+            // precondition: `find(_Val) == find(_Keyval)` (per N4958 [flat.set.modifiers]/2)
+            // the precondition cannot guarantee `_Can_insert(_Where, _Keyval)` here.
+            _STL_ASSERT(!contains(_Keyval), "find(_Val) == end(), but find(_Keyval) != end()");
+            return pair{_Emplace_hint(_Where, _STD move(_Keyval)), true};

```

Comment:
> [[.]](https://eel.is/c++draft/flat.set#modifiers-2.sentence-1) Preconditions: The conversion from x into value_type constructs an object u, for which find(x) == find(u) is true.

I think this precondition is poorly-specified... or, does the standard intentionally decide to allow those "partially inconsistent" cases?

Reply:
What the Standard *should* say is that the transparent comparison returns results that are exactly as good as the object that will be constructed eventually. That is, I think it's undesirable to add weird code to handle "partially inconsistent" comparison results. We should instead file an LWG issue instead of trying to handle phantom cases that shouldn't occur with proper usage, especially if doing so would reduce efficiency. The whole point of transparent lookup is efficiency, to avoid having to construct real keys ahead of time.

Reply:
I've made tracking issue #4105.

Reply:
I made a mistake when reasoning about inconsistency - there is no `"partially"` inconsistent types. I thought `(2)` below is a special trait for certain types, so I coined `"partially"`; now I realize `(2)` is generally true for all types under current precondition. The situation is that:
1. When a `transparent-key` type represents different things when `comparing with key_type` (per the comparator) and when `convertig to key_type` (per the type itself), it makes logical inconsistency.
2. However, there are `always` some situations where the input will not trigger precondition violation, depending on both the content in flat_set and the input. This makes the implementation unable to conceptually reject any types.

An obvious way to show how `(2)` is true and how inconsistent types are broken:
```
// empty set:
flat_set<key_type, transparent-comparator> fs;
// will never trigger precondition violation. The set is empty, so find(anything) == end().
fs.insert(arbitary-broken transparent-key);
// the same expression will violate the precondition now, as transparent-find still == end(), but find(conversion-result) != end() now.
fs.insert(the same input as above);
```

>Preconditions: The conversion from x into value_type constructs an object u, for which find(x) == find(u) is true.

As to the precondition for transparent key, I think instead of requiring certain expressions to be true, it should be undefined behavior as long as the transparent key *logically* represents different things when comparing with `key_type` (per the comparator) and when convertig to `key_type` (per the type itself).